### ISSUE TITLE
OpenAI-Compatible vLLM Server with Weight Synchronization

### ISSUE THREAD

**Issue by BjarniHaukur** [huggingface/trl#3284]
### Feature request

I propose adding a new OpenAI-compatible vLLM API server for use with the `GRPOTrainer`.

The implementation mirrors the weight syncing logic from `trl/scripts/vllm_serve.py`, but offloads most complexity to the existing `vllm.entrypoints.openai.api_server` infrastructure.

This enables training on significantly more complex rollouts than the standard synchronous .generate() endpoint can support. By supporting the OpenAI API interface, it also allows seamless integration with a wide range of existing agent frameworks and products.

This direction is a step toward reproducing pipelines like [OpenHands LM 32B](https://www.all-hands.dev/blog/introducing-openhands-lm-32b----a-strong-open-coding-agent-model). I strongly suspect that Claude 3.7 Sonnet was trained in a similar fashion, iteratively reinforced using rollouts generated through its own [Claude Code](https://docs.anthropic.com/en/docs/welcome) scaffolding.

### Motivation

Currently, TRL only supports synchronous, batched `.generate()` calls for inference. This restricts the types of rollouts that can be created,  especially in domains that benefit from having multi-step approaches, tool use, or environment interaction.

I’ve been using TRL for my Master’s thesis on reinforcement learning for language models in the program repair domain. In several GRPO experiments, I repeatedly encountered the same limitation: with `.generate()`, all context construction, planning, and feedback extraction must happen within a single call. For example, in tasks from [SWE-Gym](https://huggingface.co/SWE-Gym), the model needs to generate code edits for real repositories. To do this in one `.generate()` call, the user must manually construct the relevant repo context and later parse outputs like diffs to extract useful reward signals. This makes experimentation slow and always feels like “reinventing the wheel.”

Rather than building ad-hoc scaffolding from scratch, I began exploring how to integrate existing coding agents like [Aider](https://github.com/Aider-AI/aider) directly into the training loop. These agents already support rich workflows such as repo mapping, diff parsing, and iterative interaction—and they use the OpenAI API interface. Enabling TRL to train models through this interface would allow us to run them *in situ*, inside the same environment they’re meant to be deployed in.

This proposal aims to bridge that gap and enable more realistic, multi-step training workflows as a first-class feature in TRL.

### Your contribution

I have already developed an initial working implementation in this PR: <INSERT PR>, which introduces `vllm_serve_openai_compatible.py`.

I intend to wrap up remaining loose ends and properly test this approach, both for functional correctness and throughput benchmarking.

The draft PR also includes a few project-specific utilities (WIP) to illustrate how this can be used in practice. For example, it shows how to parallelize existing Aider instances that interact with this server to generate training data.

One open issue is how to reliably access full conversation histories for each rollout. Since API calls happen internally within the agent, we cannot assume access to `.get_conversation_history()` or similar. A possible approach is to record all requests and responses server-side and map them back to the original prompt to reconstruct complete rollouts to train on.

I’d be happy to align the implementation with TRL’s design goals and iterate toward something mergeable.

**Comment by qgallouedec** [comment]
Great job on all the work you've done, and thanks for sharing it with the TRL community!

This likely involves significant changes to TRL, but your motivations seem solid and well thought out—it feels like the right time to explore this direction.

Would it be possible to break your work into several smaller PRs? That would make the review process much smoother. For example, you could start with a PR focused on leveraging vLLM server, followed by another that integrates the tools/agents. (Of course, feel free to divide it differently if you think there's a better approach.)

**Comment by qgallouedec** [comment]
> I believe I’ve found a clean abstraction that minimizes the impact on existing code. Specifically, I’m exploring repurposing vllm_client to become a proper client interface, rather than just a wrapper around vllm_serve.py. With this in place, users could extend it and implement their own .generate() as needed.

Ok, modifying the client-server seems acceptable to me, especially if it can allow easier customization.
I look forward to hearing about your progress, keep us posted!

**Comment by kwanUm** [comment]
Hey @BjarniHaukur – I’m also looking at migrating GRPO roll-outs to an online vLLM setup for better performance and agent-style usability.
Also, really like the idea of letting the server keep the full conversation so the client doesn’t have to resend context; we’re doing that today in a separate RolloutManager on the client side, but server-side seems cleaner.
Do you already have a branch/PR that swaps the current LLM() offline path for AsyncLLMEngine? Have you been able to test throughput or latency compared to the batched .generate() route?
Would love to follow the work and possibly help test / contribute.

**Comment by BjarniHaukur** [comment]
Hey @kwanUm, still working on it. I've been super close to finishing this for a while now. The main problem resides in the online weight syncing behavior for `AsyncLLM`. I've tried a bunch of things out but the system ends up in a deadlock somewhere internally in vLLM. That isn't too surprising, vLLM only just recently merged a PR adding `collective_rpc_async`, which is the method we would need to use to initialize the updates. I'm also checking out SGLang as an async rollout client. That seems a bit more mature but still a WIP on my end.

I've put quite a lot of though into how it would be best to integrate custom clients, and I'm relatively convinced of my approach. It decouples all the generation logic from the GRPOTrainer, and offloads it to client.generate(). It receives the `inputs` dictionary containing everything from your HF dataset and returns `prompts`/`completions` + anything the user wants to pass to the reward functions.

(overly simplified example)
```python
class GenerationResult(TypedDict, total=False):
    """GRPO payload with required prompt/completion; extras allowed."""
    # Shared inputs across N rollouts in GRPO (across many GenerationResults)
    prompt: list[dict[str, str]]  # {role: str, content: str}
    # This comes after that, N different rollouts of the same prompt
    completion: list[dict[str, str]]  # {role: str, content: str}
    # Extra keys and values are forwarded to the user specified reward functions


class VLLMClient(ABC):
    @abstractmethod
    def generate(self, data: list[dict], **kwargs) -> list[GenerationResult]:
        pass

# Inside GRPOTrainer
...
output = client.generate(inputs)
...
rewards = reward_func(**output)
```

You can check out my working branch (though cautionary warning its not stable / ready at all), it might help you.
https://github.com/ASSERT-KTH/trl/tree/dev/trl
Here’s a cleaned-up, natural continuation in your original style:

I'll post here again when I have something more concrete. Would love some help in integrating this type of behavior into TRL though! There's some semblance of it in `verl`, `openpipe/art`, and `verifiers`, but nothing that quite checks all the boxes. Most existing approaches miss one of the following: full OpenAI compatibility (true async multi-step, support tool-calling etc.), or just general ease of use.

**Comment by BjarniHaukur** [comment]
Hey @qgallouedec

Finally got it working and found an abstraction that I believe could fit in TRL (#3469).

The new `vllm_serve_async.py` script behaves exactly like `vllm_serve.py`. It supports tensor and data parallelism but runs into the same error you mentioned here: [vllm-project/vllm#17079](https://github.com/vllm-project/vllm/issues/17079).

Instead of extending `VLLMClient`, I opted for adding a `rollout_func` to the `GRPOTrainer`. This gives users full control over how rollouts are generated and enables users to forward reward signals that are not present in the completions (e.g. a program's runtime, test coverage, or other environment-based feedback).

When `vllm_mode` is set to `"async_server"`, the server that is launched exposes a fully featured `/v1/` OpenAI-compatible endpoint with tool calling support. All endpoint complexity is offloaded to vLLM’s upstream implementation; the script just mirrors their design and adds our weight syncing logic.

```bash
trl vllm-serve-async \
  --model Qwen/Qwen3-8B \
  --max_model_len 8192 \
  --enable-auto-tool-choice \
  --reasoning_parser deepseek_r1 \
  --tool-call-parser hermes
```

This allows any LLM-powered application with measurable reward metrics to be trained with `GRPOTrainer` by minimally wrapping the app in a `rollout_func` which interacts with this server.

My [CodeRepairRL](https://github.com/ASSERT-KTH/CodeRepairRL) project provides an example of a [rollout_func](https://github.com/ASSERT-KTH/CodeRepairRL/blob/master/src/agents/nano_agent.py) using a terminal-based coding agent ([Nano-Agent](https://github.com/ASSERT-KTH/nano-agent)).

@kwanUm, this might be of interest to you too!

**Comment by qgallouedec** [comment]
Hey @BjarniHaukur, sorry for the very slow reply here, we're catching up on the backlog.

Good news: the direction you proposed has since landed on `main`. `GRPOTrainer` now takes an experimental `rollout_func`, that hands you the per-process prompt slice and lets you return `prompt_ids`/`completion_ids`/`logprobs` plus any extra fields, which are forwarded to the reward functions (essentially the abstraction you sketched in #3469).
For the async/agent side there is now `trl.experimental.async_grpo` (an `AsyncGRPOTrainer` that streams rollouts from a vLLM server and handles online weight sync) and `trl.experimental.external_harness_grpo`, which puts an OpenAI/Anthropic-compatible server in front of the policy so an unmodified coding agent can drive rollouts with token-exact capture (see https://github.com/huggingface/trl/blob/e7ffa913e507c23aa0e7747f69bc1908977eeb18/trl/experimental/external_harness_grpo/README.md) — very close to the Aider-in-the-loop workflow you described.
